src/foam/box/SocketService.js

```diff
@@ -0,0 +1,183 @@
+ */
+
+foam.CLASS({
+  package: 'foam.box',
+  name: 'SocketService',
+  flags: ['java'],
+
+  javaImports: [
+    'foam.core.*',
+    'foam.nanos.box.NanoServiceRouter',
+    'foam.nanos.logger.Logger',
+    'foam.lib.json.JSONParser',
+    'foam.lib.json.Outputter',
+    'java.net.ServerSocket',
+    'java.util.Random',
+    'java.util.Base64',
+    'java.net.Socket',
+    'java.io.IOException',
+    'java.io.ByteArrayOutputStream',
+    'foam.box.Message'
+  ],
+
+  properties: [
+    {
+      class: 'Boolean',
+      name: 'listen',
+      value: true
+    },
+    {
+      class: 'Boolean',
+      name: 'listening'
+    },
+    {
+      class: 'Int',
+      name: 'port_',
+      value: 7000
+    },
+    {
+      class: 'Object',
+      name: 'serverSocket',
+      javaType:   'java.net.ServerSocket'
+    },
+    {
+      class: 'Object',
+      name: 'parser',
+      javaType: 'foam.lib.json.JSONParser',
+      javaFactory: `return getX().create(JSONParser.class);`
+    },
+    {
+      class: 'Object',
+      name: 'router',
+      javaType: 'foam.nanos.box.NanoServiceRouter',
+      javaFactory: `
+        return getX().create(NanoServiceRouter.class);
+      `
+    }
+  ],
+
+  constants: [
+    {
+      type: 'Integer',
+      name: 'BUFFERSIZE',
+      value: 1024
+    },
+  ],
+
+  methods: [
+    {
+      name: 'setupServer',
+      args: [
+        {
+          name: 'port',
+          type: 'Integer'
+        }
+      ],
+      javaCode:
+        `try {
+          ServerSocket serverSocket = new ServerSocket(port);
+          setServerSocket(serverSocket); 
+          setPort_(port);
+          while (true) {
+            new SocketClientHandler(getX(), getServerSocket().accept(), getParser(), getRouter()).start();
+          }
+        } catch (IOException e) {
+          port = 10000 + (new Random().nextInt(10000) + 1);
+          setupServer(port);
+        }`
+    },
+    {
+      name: 'stop',
+      javaCode:
+        `try {
+          getServerSocket().close();
+        } catch (IOException ex) {
+          ((Logger) getX().get("logger")).error(ex);
+        }`     
+    }
+  ],
+
+  axioms: [
+    {
+      name: 'javaExtras',
+      buildJavaClass: function(cls) {
+        cls.extras.push(foam.java.Code.create({
+          data:
+            `
+            private static class SocketClientHandler extends Thread {
+              private X x;
+              private Socket clientSocket;
+              private JSONParser parser;
+              private NanoServiceRouter router;
+          
+              public SocketClientHandler(X x, Socket socket, JSONParser parser, NanoServiceRouter router) {
+                this.x = x;
+                this.parser = parser;
+                this.clientSocket = socket;
+                this.router = router;
+              }
+
+              public void run() {
+                try {
+                  String mStr = Base64.getEncoder().encodeToString(read(this.clientSocket));
```

Comment:
No need to use Base64 over a raw socket.  Base64 is primarily used to encode binary data in a text only medium like an email/url/json etc.

We can just send raw bytes over the wire which saves spaces and time.